### PR TITLE
Separate self-attraction and loading as a new module

### PR DESCRIPTION
**Pull request by herrwang0**
Calculation for self-attraction and loading (SAL) is taken out from `MOM_tidal_forcing` and housed in a new module (`MOM_self_attr_load`) as SAL is a process not limited to tides.

The new module includes both online spherical harmonics method and scalar approximation. Read-in method (TIDAL_SAL_FROM_FILE) is kept in the tidal forcing module as it is specific to tides. For the iterative method (USE_PREV_TIDES), the updating part that is tied to the scalar approximation is moved to the new SAL model, while the read-in part remains in the tidal_forcing module.

Module `MOM_tidal_forcing` is designated only to calculations of the astronomical forcing and periodic tidal SAL that is based on previous solutions from input files. This makes the tidal forcing module now only contain calculations independent from the ocean's state and the only input variable it needs is the current time.

* A new parameter `CALCULATE_SAL` is added to control if SAL is calculated. At the moment, the default is linked to if `TIDES` is used.
* A bug with SSH used for SAL with flooding points are now fixed.
* New diagnostics on the SAL, equilibrium tides and tidal SAL are added.
* Input parameters are renamed to remove 'TIDAL' or 'TIDES' from the SAL fields. This is done in a non-disruptive way. Old parameters with scalar SAL are still accepted but with a warning message.

## Answer changes
Due to the rearrangement of the summations, answers are changed at bit level.
* For FV Pressure Force in Boussinesq mode, a `TIDES_ANSWER_DATE` runtime flag is added to restore old answers. Any integer > 20230701 gives the new answer. The current default is 20230630.
* For the other options, FV Pressure in non-Boussinesq and Montgomery Pressure Force, which presumably are rarely used for tides and not included in the testing suites, answers are expected to change.

### CODE REVIEW

**Review by Hallberg-NOAA — COMMENTED** [review]
I have read through all of these changes, and they all make sense to me.  I am tempted to approve it right now as-is, but for one minor procedural issue.

   This PR is bending the usual rules about how we handle obsoleting (or revising) parameter names only after consortium-wide consultation. At this point I suspect there are not yet too many other groups in the MOM6 consortium using the tidal specification parameters for on-line SAL, so we are probably OK with this, but I am going to suggest that we defer acting on this PR until after it has been discussed.   @herrwang0, could I ask you to please add the proposed parameter name and interface changes to the shared google drive document that we will be using to discuss parameter name and default changes (ask me directly if you need help finding it), and we will discuss it next Monday (July 17, 2023).  I strongly suspect that we are going to find widespread alignment on these changes, but I don't want anyone to be surprised by them.

**Comment by herrwang0** [comment]
@Hallberg-NOAA This is a very good point. I agree the annual parameter review is the perfect opportunity to officially change these parameter names (and defaults). It would probably also simplify the code by a lot. I will add these proposed changes in the shared document.

**Review by Hallberg-NOAA — CHANGES_REQUESTED** [review]
This PR is very close to being ready to go, but please revise it as suggested during the discussion of these changes during the consortium-wide MOM6 dev call on July 17th.  Specifically, we should defer the answer-changing default setting of the newly added ANSWER_DATE parameter, and please reexamine the new parameter names to see whether there is a more self-consistent naming scheme that might make sense.

File: src/parameterizations/lateral/MOM_self_attr_load.F90
Comment (Hallberg-NOAA) [inline src/parameterizations/lateral/MOM_self_attr_load.F90:196]: Please correct the indenting on this line.  As written, the indenting makes it look like this is inside of the `if (tide_sal_scalar_value /= 0.0)` block.

**Review by Hallberg-NOAA — CHANGES_REQUESTED** [review]
This is getting very close, but I have added a few more specific comments about consistency between the new the get_param calls that will need to be addressed before this can be merged in.

File: src/parameterizations/lateral/MOM_tidal_forcing.F90
Comment (Hallberg-NOAA) [inline src/parameterizations/lateral/MOM_tidal_forcing.F90:358]: The logic for setting SAL_SCALAR_VALUE should probably be revised to mirror what is done in SAL_init, although the fail_if_missing flag would have to be removed if we are setting a default value, as is done there.

File: src/parameterizations/lateral/MOM_self_attr_load.F90
Comment (herrwang0) [reply src/parameterizations/lateral/MOM_self_attr_load.F90:206]: > Because of the do_not_log arguments, these 3 get_param calls can be moved outside of the `if (CS%use_sal_sht) then` test.

If I understand correctly. Without the `if` test, these parameters wouldn't be logged when the switch is off, but also wouldn't be marked as "unused". Is it going to be a problem?

Comment (Hallberg-NOAA) [reply src/parameterizations/lateral/MOM_self_attr_load.F90:206]: Your understanding is correct.  The question comes down to how we want to handle the case where the SAL is used and one of these otherwise valid parameters is in an input file (perhaps because of shared input files across sensitivity experiments).  My preference is to allow this, but others (e.g. Alistair) would disagree from time to time.  Perhaps it all comes down to your most recent cause for an annoying model failure.
It is your call on this one, but we don't need both the do_not_log argument and for the calls to be inside of the if block, because as written they are duplicative, if benign.

**Review by Hallberg-NOAA — APPROVED** [review]
Thank you for these final minor modifications.  I think that this PR is now ready to go (assuming, of course, that it will pass the regression testing).

**Review by Hallberg-NOAA — CHANGES_REQUESTED** [review]
Unfortunately, this PR is changing answers for some cases with the tides enabled, such as nonBous_global and tides_025.  Almost certainly the issue is that we have changed order of arithmetic, but I suspect that we will need to track this down and understand it thoroughly before we can merge this PR in to MOM6.

**Comment by Hallberg-NOAA** [comment]
I have gone back and checked and found that I was in error with my previous comment.   (I am sorry about the confusion my previous misstatement created, @herrwang0 !)  The global test case in MOM6_examples _does_ reproduce with this PR.  It is only nonBous_global and tides_025 where the answers changed, and as noted above, these are Non-Boussinesq test cases.

  However, when I run the tides_025 case with the added line `#override BOUSSINESQ = True` in the MOM_override file, the answers do reproduce with the old and new versions of the code.

**Comment by Hallberg-NOAA** [comment]
If you don't mind, @herrwang0, I think that I am going to try adding the TIDES_ANSWER_DATE code to the semi-Boussinesq cases as well so that we reproduce the answers for these other two test cases.  I will then send those (hopefully minor) changes back as  PR to your branch.

**Comment by Hallberg-NOAA** [comment]
@herrwang0, I have added a PR onto your feature branch for this PR at github.com/herrwang0/MOM6/pull/1 which is allowing the non-Boussinesq test cases in the MOM6-examples suite to also reproduce the existing answers with an appropriate value of TIDES_ANSWER_DATE.  If you agree with these changes, please merge them into your branch, and then we should be able to merge this PR into dev/gfdl.

**Comment by herrwang0** [comment]
@Hallberg-NOAA Thanks for making the patch for preserving the non Boussinesq answers. While we are at this, shall we follow suits and clean up the Boussinesq counterpart? (I coded it in a very redundant way, with the expectation that we could simply remove the else branch some time in the future. But now I think it is probably better with your structure). I can do a quick test and submit a patch later tonight.

**Review by Hallberg-NOAA — APPROVED** [review]

**Comment by marshallward** [comment]
https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/20415 :heavy_check_mark: :yellow_circle:

**Comment by marshallward** [comment]
Some of the commits seem redundant, but since they are all well-commented I will rebase them all into the repo.

**Pull request merged**
